Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs

```diff
@@ -0,0 +1,267 @@
+                items.Add(f);
+            }
+
+            // Assert
+            items.Last().Should().Be(@"d:\test1\image2.png");
+        }
+
+        [Fact]
+        public void Search_FilterTxtFilesAndSkipFilteredFileByFilteredFileFoundEvent_ShouldFilterTxtFilesAndNotReturnSkipedFile()
+        {
+            // Arrange
+            var fileSystem = new MockFileSystem(new Dictionary<string, MockFileData>
+            {
+                { @"d:\test1\tes1_1\text.txt", new MockFileData("Text") },
+                { @"d:\test1\tes1_2\text2.txt", new MockFileData("Text2") },
+                { @"d:\test1\im.gif", new MockFileData(System.Array.Empty<byte>()) }
+            });
+
+            Func<FileSystemInfo, bool> filter = (FileSystemInfo e) => e.Extension == ".txt";
+
+            // Act
+            var visitor = new FileSystemVisitor(fileSystem, filter);
+
+            visitor.FilteredFileFound += (FileSystemInfo e) =>
+            {
+                if (e.Name.Contains("im"))
+                {
+                    return Actions.Skip;
+                }
+                else
+                {
+                    return Actions.Continue;
+                }
+            };
+
+            var items = visitor.Search(@"d:\test1");
+
+            // Assert
+            items.Should().BeEquivalentTo(@"d:\test1\tes1_1", @"d:\test1\tes1_2");
+        }
+
+        [Fact]
+        public void Search_FilterFilesAndStopOnSearchedDirectoryByFilteredDirectoryFoundEvent_FilesAreFilteredAndLastItemIsSearchedDirectory()
+        {
+            // Arrange
+            var items = new List<string>();
+
+            var fileSystem = new MockFileSystem(new Dictionary<string, MockFileData>
+            {
+                { @"d:\test1\test1_1\text.txt", new MockFileData("Text") },
+                { @"d:\test1\test1_2\text2.txt", new MockFileData("Text2") },
+                { @"d:\test1\im.gif", new MockFileData(System.Array.Empty<byte>()) }
+            });
+
+            fileSystem.AddDirectory(@"d:\test1\dir1");
+            fileSystem.AddDirectory(@"d:\test1\dir2");
+            fileSystem.AddDirectory(@"d:\test1\test1_1\dir3");
+            fileSystem.AddDirectory(@"d:\test1\test1_1\dir4");
+
+            Func<FileSystemInfo, bool> filter = (FileSystemInfo e) => e.Name.Contains("test");
+
+            // Act
+            var visitor = new FileSystemVisitor(fileSystem, filter);
+
+            visitor.FilteredDirectoryFound += (FileSystemInfo e) =>
+            {
+                if (e.Name.Contains("dir3"))
+                {
+                    return Actions.Stop;
+                }
+                else
+                {
+                    return Actions.Continue;
+                }
+            };
+
+            foreach (var f in visitor.Search(@"d:\test1"))
+            {
+                items.Add(f);
+            }
+
+            // Assert
+            items.Select(i => i.Substring(i.LastIndexOf('\\'))).Should().NotContain("test");
```

Comment:
You're reinventing the wheel here 😃 

```suggestion
            items.Select(info => new DirectoryInfo(info).Name).Should().NotContain("test");
```